url284ba87ea1_swabhs_forbidden_paths_tgz-pJ8-tagging_hypergraph_TaggingHypergraphGeneratorJ8/src/learning/InputReader.java

```diff
@@ -0,0 +1,83 @@
+package learning;
+
+import java.io.BufferedReader;
+import java.io.DataInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+
+public class InputReader {
+	
+	/**
+	 * Reads the initial hyperedge weights from a file
+	 * @param trigramWeights
+	 * @param tagWeights
+	 */
+	public static Map<String, Double> readWeights(File weightsFile) {
+		Map<String, Double> weights = new TreeMap<String, Double>();
+		try {						
+			FileInputStream fstream = new FileInputStream(weightsFile);
+			DataInputStream in = new DataInputStream(fstream);
+			BufferedReader br = new BufferedReader(new InputStreamReader(in));
+			String strLine;
+			
+			while ((strLine = br.readLine()) != null) {
+				
+				String[] items = strLine.split(" ");
+				weights.put(items[0], Double.parseDouble(items[1]));
+				
+			}				
+		} catch(IOException e) {
+			e.printStackTrace();
+		}
+		
+		return weights;
+	}
+	
+	/** 
+	 * 
+	 * @param trainingFile
+	 * @param weightsMap
+	 * @return
+	 */
+	public static List<LearningExample> readExample(
+			File trainingFile, Map<String, Double> weightsMap) {
+		
+		List<LearningExample> examples = new ArrayList<LearningExample>();
+		try {
+			
+			FileInputStream fstream = new FileInputStream(trainingFile);
+			DataInputStream in = new DataInputStream(fstream);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *in*. Execution paths that do not contain closure statements were detected. Ensure *in* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).